File: scripts/probe-flowchart.py
#!/usr/bin/env python3
"""Run the generated `coda_flow_layers` and check it answers what the canvas answered.

    pnpm probe:flowchart

The second third of a three-language check; `scripts/probe-flowchart.ts` writes the reference.
The helper is read out of the golden notebook rather than transcribed here, so what runs is
exactly what the exporter writes — `probe-py-helpers.py`' arrangement and for its reason.

It also runs `multipartite_layout` over the result, because that is the call the emitted cell
makes and it is the one that can reject a layering rather than merely disagree with it: the
layout requires the subset key on every node and raises otherwise, which a layering that skipped
an isolated node would trip.
"""
import json
import sys
from pathlib import Path

import networkx as nx

ROOT = Path(__file__).resolve().parent.parent
NOTEBOOK = ROOT / "src" / "export" / "python" / "__fixtures__" / "everything.ipynb"
REFERENCE = Path(sys.argv[1] if len(sys.argv) > 1 else "/tmp/coda-flowchart-probe.json")


def load_source(notebook: Path, marker: str) -> str:
    """The generated cell containing `marker`, as text."""
    book = json.loads(notebook.read_text())
    for cell in book["cells"]:
        source = "".join(cell.get("source", []))
        if cell["cell_type"] == "code" and marker in source:
            return source
    raise SystemExit(f"no cell containing {marker!r} in {notebook}")


def load_cell(notebook: Path, marker: str, ns: dict) -> dict:
    """Exec that cell into `ns`, and hand `ns` back."""
    exec(compile(load_source(notebook, marker), str(notebook), "exec"), ns)
    return ns


fails = []


def check(name, cond, detail=""):
    print(f"{'ok  ' if cond else 'FAIL'} {name}{f'   {detail}' if detail else ''}")
    if not cond:
        fails.append(name)


if not REFERENCE.exists():
    raise SystemExit(f"{REFERENCE} missing — run `pnpm probe:flowchart` (the vite-node half) first")

ns = load_cell(NOTEBOOK, "def coda_flow_layers(", {})
coda_flow_layers = ns["coda_flow_layers"]

for case in json.loads(REFERENCE.read_text()):
    graph = nx.DiGraph()
    for node in case["nodes"]:
        graph.add_node(node, hops=case["hops"][node])
    for source, target, weight in case["edges"]:
        graph.add_edge(source, target, weight=weight)

    layers = coda_flow_layers(
        graph, **({"layer_attr": case["layerColumn"]} if case["layerColumn"] else {})
    )
    expected = case["layers"]
    same = all(layers.get(node) == expected[node] for node in case["nodes"])
    check(
        f"{case['name']}: the same layering as the canvas",
        same,
        "" if same else f"{ {n: layers.get(n) for n in case['nodes']} } vs {expected}",
    )

    # The feedback classification falls out of the layering, so a layering that agrees and a
    # classification that does not would mean the two disagree about which edges exist.
    kinds = {}
    for source, target, _ in case["edges"]:
        if source == target:
            kinds[f"{source}->{target}"] = "self"
        elif layers[target] > layers[source]:
            kinds[f"{source}->{target}"] = "forward"
        elif layers[target] < layers[source]:
            kinds[f"{source}->{target}"] = "back"
        else:
            kinds[f"{source}->{target}"] = "within"
    check(f"{case['name']}: the same feedback edges", kinds == case["kinds"],
          "" if kinds == case["kinds"] else f"{kinds} vs {case['kinds']}")

    # The call the emitted cell actually makes. It raises on a node with no subset key, which is
    # the failure a layering that quietly skipped the isolated node would produce.
    nx.set_node_attributes(graph, layers, "coda_layer")
    pos = nx.multipartite_layout(graph, subset_key="coda_layer", align="vertical")
    check(f"{case['name']}: multipartite_layout places every node", len(pos) == len(case["nodes"]))

    # And that the layering is what decides the axis the layers run along — the property the
    # figure rests on, rather than the numbers themselves.
    by_layer = {}
    for node, at in pos.items():
        by_layer.setdefault(layers[node], set()).add(round(float(at[0]), 6))
    one_x_per_layer = all(len(xs) == 1 for xs in by_layer.values())
    ordered = [x for _, xs in sorted(by_layer.items()) for x in xs]
    check(
        f"{case['name']}: one column per layer, in layer order",
        one_x_per_layer and ordered == sorted(ordered),
        "" if one_x_per_layer else f"{by_layer}",
    )

# ---------------------------------------------------------------------------
# coda_paths_network, the other helper the Flow Chart depends on
# ---------------------------------------------------------------------------

# `neuron.paths` had to start binding its Network port once something downstream read it, and the
# builder's one interesting rule is that the shift is *within* a path. Applied to the frame whole
# it joins the last body of one route to the first body of the next — an edge that does not exist,
# carrying the next route's first weight, indistinguishable from a real one in the picture.
print()
print("-- coda_paths_network --")

import pandas as pd

pn = load_cell(NOTEBOOK, "def coda_paths_network(", {"pd": pd, "nx": nx})
coda_paths_network = pn["coda_paths_network"]

# `fetch_paths`' documented shape: one row per step, `weight` the strength from the previous body
# in the same path, 0 on a route's first row. Two routes, deliberately arranged so the last body
# of path 0 and the first body of path 1 are different neurons — which is what makes the spurious
# edge visible at all.
frame = pd.DataFrame(
    [
        {"path": 0, "bodyId": 1, "type": "L1", "weight": 0},
        {"path": 0, "bodyId": 2, "type": "Tm3", "weight": 11},
        {"path": 0, "bodyId": 3, "type": "T4a", "weight": 15},
        {"path": 1, "bodyId": 4, "type": "L2", "weight": 0},
        {"path": 1, "bodyId": 2, "type": "Tm3", "weight": 7},
        {"path": 1, "bodyId": 3, "type": "T4a", "weight": 15},
    ]
)
graph = coda_paths_network(frame)

# `check` takes the name first. Written the other way round every one of these passed on a
# truthy string while asserting nothing — which is what the printed labels said, and is the
# reason to read a probe's first run rather than only its exit code.
check(
    f"a node per body ({graph.number_of_nodes()})",
    set(graph.nodes()) == {"1", "2", "3", "4"},
)
check(
    "an edge per consecutive pair within a path",
    set(graph.edges()) == {("1", "2"), ("2", "3"), ("4", "2")},
    str(sorted(graph.edges())),
)
check(
    "and no edge from one route's last body to the next route's first",
    not graph.has_edge("3", "4"),
)
# `2 -> 3` is on both routes at the same weight, so it is one edge at that weight rather than 30.
check(
    "a connection on two routes is one edge at its own weight, not the sum",
    graph.edges["2", "3"]["weight"] == 15,
    str(graph.edges["2", "3"]["weight"]),
)
check("and each node carries its type", graph.nodes["2"]["type"] == "Tm3")
check("an empty frame is an empty graph", coda_paths_network(frame.iloc[0:0]).number_of_nodes() == 0)

# ---------------------------------------------------------------------------
# The emitted cell itself, run
# ---------------------------------------------------------------------------

# Lifted out of the golden notebook and executed against a graph of our own, because the checks
# above cover the *helper* and the golden covers the *text*, and neither runs the twenty lines
# that draw. Its R twin found four bugs in exactly those lines.
print()
print("-- the emitted cell, run --")

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt


# Carries everything the cell reads: a `hop` column to layer by, a `role` column to label by,
# `weight` for the widths and `pairs` for the labels — and an edge skipping a layer, which is
# where a straight-line renderer and Coda's routed one visibly part company.
paths_network = nx.DiGraph()
for name, hop, role in [
    ("in1", 0, "source"),
    ("via", 1, "via"),
    ("via2", 2, "via"),
    ("out", 3, "target"),
]:
    paths_network.add_node(name, hop=hop, role=role)
for source, target, weight, pairs in [
    ("in1", "via", 120, 3),
    ("via", "via2", 45, 2),
    ("via2", "out", 300, 7),
    ("in1", "out", 8, 1),
]:
    paths_network.add_edge(source, target, weight=weight, pairs=pairs)

cell_ns = {
    "nx": nx,
    "pd": pd,
    "plt": plt,
    "coda_flow_layers": coda_flow_layers,
    "paths_network": paths_network,
}
try:
    exec(compile(load_source(NOTEBOOK, "coda_flow_layers(_figure"), "<cell>", "exec"), cell_ns)
    ran, why = True, ""
except Exception as error:  # noqa: BLE001 — the probe's whole job is to report this
    ran, why = False, f"{type(error).__name__}: {error}"
plt.close("all")

check("the emitted cell runs", ran, why)
check(
    "it binds the pass-through and the selection",
    "flow_chart_out" in cell_ns and "flow_chart_selected" in cell_ns,
)
if "flow_chart_selected" in cell_ns:
    # The fixture's selection names a neuron this graph does not have, so the frame is empty —
    # what is checked is that the column survives, since a frame with no columns at all is what
    # makes a later merge against a real one fail.
    selected = cell_ns["flow_chart_selected"]
    check(
        "the selection frame keeps a neuronId column",
        "neuronId" in selected.columns,
        f"{len(selected)} rows",
    )

print()
print(f"{len(fails)} failed" if fails else "all passed")
sys.exit(1 if fails else 0)

File: scripts/probe-flowchart.R
#!/usr/bin/env Rscript
# Run the generated `coda_flow_layers` and check it answers what the canvas answered.
#
#     pnpm probe:flowchart
#
# The last third of a three-language check; `scripts/probe-flowchart.ts` writes the reference.
# The helper is read out of the golden R Markdown rather than transcribed here, so what runs is
# exactly what the exporter writes — `probe-r-helpers.R`' arrangement and for its reason.
#
# **The second half is the part that could not be settled by reading.** The emitted cell draws on
# `layout_with_sugiyama`'s `extd_graph`, which is the original graph plus a dummy vertex per layer
# a long edge crosses — that is the whole reason this emitter reaches for Sugiyama rather than
# `layout_as_tree`, since it is what routes an arrow around the boxes between its ends. But an
# edge that gets split is not the edge that arrived, and whether the pieces keep the original's
# attributes decides whether `E(.g)$weight` and `E(.g)$label` are the right numbers, the wrong
# numbers, or NULL. igraph's documentation does not say. So it is measured here.

suppressPackageStartupMessages(library(igraph))

root <- normalizePath(file.path(dirname(sub("^--file=", "", grep("^--file=",
  commandArgs(trailingOnly = FALSE), value = TRUE)[1])), ".."))
rmd <- file.path(root, "src", "export", "r", "__fixtures__", "everything.Rmd")
reference <- if (length(commandArgs(trailingOnly = TRUE))) {
  commandArgs(trailingOnly = TRUE)[1]
} else "/tmp/coda-flowchart-probe.json"

if (!file.exists(reference)) {
  stop(sprintf("%s missing - run the vite-node half of `pnpm probe:flowchart` first", reference))
}

# The generated helper, lifted out of the golden document by its own first line.
lines <- readLines(rmd, warn = FALSE)
start <- grep("^coda_flow_layers <- function", lines)[1]
if (is.na(start)) stop("no coda_flow_layers in the golden document")
# To the closing brace at column zero, which is where a top-level function ends.
ends <- grep("^\\}$", lines)
stop_at <- ends[ends > start][1]
eval(parse(text = paste(lines[start:stop_at], collapse = "\n")), envir = globalenv())

fails <- character(0)
check <- function(name, cond, detail = "") {
  cat(sprintf("%s %s%s\n", if (isTRUE(cond)) "ok  " else "FAIL", name,
              if (nzchar(detail)) paste0("   ", detail) else ""))
  if (!isTRUE(cond)) fails <<- c(fails, name)
}

# `simplifyVector = FALSE` throughout, so an edge stays a three-element list rather than being
# collapsed into a character matrix — which is what jsonlite does by default to an array of
# same-length arrays, and which silently turns `e[[2]]` into the wrong subscript.
cases <- jsonlite::fromJSON(reference, simplifyVector = FALSE)

for (case in cases) {
  ids <- unlist(case$nodes)
  edges <- do.call(rbind, lapply(case$edges, function(e) {
    data.frame(from = e[[1]], to = e[[2]], weight = as.numeric(e[[3]]))
  }))
  g <- igraph::graph_from_data_frame(edges, directed = TRUE, vertices = data.frame(
    name = ids,
    hops = vapply(ids, function(id) {
      v <- case$hops[[id]]
      if (is.null(v)) NA_real_ else as.numeric(v)
    }, numeric(1))
  ))

  layers <- if (is.null(case$layerColumn)) {
    coda_flow_layers(g)
  } else {
    coda_flow_layers(g, layer_attr = case$layerColumn)
  }
  # Coda is 0-based and igraph's `layers` argument is 1-based, so the helper returns 1-based and
  # the comparison takes the offset back off. That offset is real rather than cosmetic — see the
  # helper — so it is stated here rather than hidden in either side.
  got <- setNames(as.integer(layers) - 1L, igraph::V(g)$name)
  want <- vapply(ids, function(id) as.integer(case$layers[[id]]), integer(1))
  same <- all(got[ids] == want)
  check(sprintf("%s: the same layering as the canvas", case$name), same,
        if (same) "" else paste(sprintf("%s=%d/%d", ids, got[ids], want), collapse = " "))

  kinds <- vapply(case$edges, function(e) {
    a <- got[[e[[1]]]]; b <- got[[e[[2]]]]
    if (e[[1]] == e[[2]]) "self" else if (b > a) "forward" else if (b < a) "back" else "within"
  }, character(1))
  names(kinds) <- vapply(case$edges, function(e) paste0(e[[1]], "->", e[[2]]), character(1))
  want_kinds <- unlist(case$kinds)
  check(sprintf("%s: the same feedback edges", case$name),
        identical(kinds[names(want_kinds)], want_kinds))

  # The call the emitted cell makes.
  sugi <- igraph::layout_with_sugiyama(g, layers = as.integer(layers))
  check(sprintf("%s: sugiyama places every real vertex", case$name),
        nrow(sugi$layout) == length(ids))

  # One row per layer, and the layers run **downwards**: layer 1 sits at the largest y, which
  # plots at the top since igraph's y increases upward. Measured, not assumed — the first
  # version of the emitted cell exchanged the two axes for left-to-right and drew the circuit
  # right to left, which is what this check now pins.
  ys <- split(round(sugi$layout[, 2], 6), got[igraph::V(g)$name])
  one_per_layer <- all(vapply(ys, function(v) length(unique(v)) == 1, logical(1)))
  descending <- unlist(lapply(ys[order(as.integer(names(ys)))], function(v) unique(v)))
  check(sprintf("%s: one row per layer, layers running downwards", case$name),
        one_per_layer && !is.unsorted(rev(descending)),
        if (one_per_layer) "" else paste(capture.output(str(ys)), collapse = " "))
}

# ---------------------------------------------------------------------------
# What `extd_graph` does to an edge it splits
# ---------------------------------------------------------------------------

cat("\n-- extd_graph, measured --\n")

# `a -> d` skips two layers, so Sugiyama has to route it: this is the case the emitter exists for.
skip <- igraph::graph_from_data_frame(
  data.frame(
    from = c("a", "b", "c", "a"),
    to = c("b", "c", "d", "d"),
    weight = c(10, 20, 30, 999)
  ),
  directed = TRUE
)
layers <- coda_flow_layers(skip)
sugi <- igraph::layout_with_sugiyama(skip, layers = as.integer(layers))
ext <- sugi$extd_graph

cat(sprintf("original: %d vertices, %d edges\n", igraph::vcount(skip), igraph::ecount(skip)))
cat(sprintf("extended: %d vertices, %d edges\n", igraph::vcount(ext), igraph::ecount(ext)))
cat(sprintf("edge attributes on the extended graph: %s\n",
            paste(igraph::edge_attr_names(ext), collapse = ", ") ))
cat(sprintf("vertex attributes on the extended graph: %s\n",
            paste(igraph::vertex_attr_names(ext), collapse = ", ") ))

check("extd_graph adds a dummy vertex per crossed layer",
      igraph::vcount(ext) > igraph::vcount(skip),
      sprintf("%d -> %d", igraph::vcount(skip), igraph::vcount(ext)))

# The question. If `weight` is absent, `E(.g)$weight` is NULL and every arrow in the emitted
# figure is one width. If it is present but the split pieces carry NA, the widths are wrong for
# exactly the long edges. Either way the emitter has to know which.
has_weight <- "weight" %in% igraph::edge_attr_names(ext)
cat(sprintf("weight kept on the extended graph: %s\n", has_weight))
check("extd_graph drops every original edge attribute", !has_weight)
check("extd_graph carries `orig`, the 1-based original edge index",
      "orig" %in% igraph::edge_attr_names(ext))

# So this is how the emitted cell has to read a weight, and the check is that it lands the
# original's value on every piece of a split edge.
orig <- igraph::E(ext)$orig
mapped <- igraph::E(skip)$weight[orig]
pieces <- which(orig == 4L)
check("reading through `orig` gives every piece of a split edge the original's weight",
      length(pieces) == 3 && all(mapped[pieces] == 999),
      sprintf("%d pieces, weights %s", length(pieces),
              paste(mapped[pieces], collapse = "/")))

# And the gift: igraph has already suppressed the arrowheads on all but the last piece, so a
# routed arrow draws one head at its real target and the emitter must not overwrite this.
modes <- igraph::E(ext)$arrow.mode[pieces]
check("arrow.mode leaves one arrowhead on the last piece of a split edge",
      sum(modes != 0) == 1 && tail(modes, 1) != 0,
      paste(modes, collapse = "/"))

# The real vertices come first, which is what makes `.real <- seq_len(vcount(g))` right.
check("the dummy vertices come after every real one",
      identical(igraph::V(ext)$dummy,
                c(rep(FALSE, igraph::vcount(skip)),
                  rep(TRUE, igraph::vcount(ext) - igraph::vcount(skip)))))

# ---------------------------------------------------------------------------
# The emitted cell itself, run
# ---------------------------------------------------------------------------

# Lifted out of the golden document and executed against a graph of our own, because the checks
# above cover the *helper* and the golden covers the *text* — and neither runs the twelve lines
# that actually draw. Both bugs this probe found were in those twelve lines, not in the helper.
cat("\n-- the emitted cell, run --\n")

suppressPackageStartupMessages({
  library(dplyr)
  library(tibble)
})

chunk_start <- grep("^```\\{r flow-chart\\}$", lines)[1]
if (is.na(chunk_start)) stop("no flow-chart chunk in the golden document")
fence <- grep("^```$", lines)
chunk_end <- fence[fence > chunk_start][1]
cell <- lines[(chunk_start + 1):(chunk_end - 1)]

# The cell's input variable, bound to a graph carrying everything it reads: a `hop` column to
# layer by, a `role` column to label by, `weight` for the widths and `pairs` for the labels —
# and an edge that skips a layer, which is the case `extd_graph` exists for.
paths_network <- igraph::graph_from_data_frame(
  data.frame(
    from = c("in1", "via", "via2", "in1"),
    to = c("via", "via2", "out", "out"),
    weight = c(120, 45, 300, 8),
    pairs = c(3L, 2L, 7L, 1L)
  ),
  directed = TRUE,
  vertices = data.frame(
    name = c("in1", "via", "via2", "out"),
    hop = c(0L, 1L, 2L, 3L),
    role = c("source", "via", "via", "target")
  )
)

pdf(NULL)
ran <- tryCatch({
  eval(parse(text = paste(cell, collapse = "\n")), envir = globalenv())
  TRUE
}, error = function(e) {
  cat(sprintf("   error: %s\n", conditionMessage(e)))
  FALSE
}, warning = function(w) {
  cat(sprintf("   warning: %s\n", conditionMessage(w)))
  FALSE
})
invisible(dev.off())

check("the emitted cell runs with no error or warning", ran)
check("it binds the pass-through and the selection",
      exists("flow_chart_out") && exists("flow_chart_selected"))
if (exists("flow_chart_selected")) {
  # The fixture's selection names a neuron this graph does not have, so the frame is empty —
  # what is checked is that the column survives, since an empty double column is what makes a
  # later `bind_rows` against a real one error outright.
  check("the selection frame keeps a character neuronId column",
        is.character(flow_chart_selected$neuronId),
        sprintf("%d rows", nrow(flow_chart_selected)))
}

if (length(fails)) {
  cat(sprintf("\n%d failed\n", length(fails)))
  quit(status = 1)
}
cat("\nall passed\n")
